// botframework-connector/lib/auth/certificateServiceClientCredentialsFactory.js
"use strict";
/**
 * @module botframework-connector
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CertificateServiceClientCredentialsFactory = void 0;
const serviceClientCredentialsFactory_1 = require("./serviceClientCredentialsFactory");
const assert_1 = require("assert");
const certificateAppCredentials_1 = require("./certificateAppCredentials");
/**
 * A Certificate implementation of the [ServiceClientCredentialsFactory](xref:botframework-connector.ServiceClientCredentialsFactory) abstract class.
 */
class CertificateServiceClientCredentialsFactory extends serviceClientCredentialsFactory_1.ServiceClientCredentialsFactory {
    /**
     * Initializes a new instance of the CertificateServiceClientCredentialsFactory class.
     *
     * @param appId Microsoft application Id related to the certificate.
     * @param certificateThumbprint A hex encoded thumbprint of the certificate.
     * @param certificatePrivateKey A PEM encoded certificate private key.
     * @param tenantId Optional. The oauth token tenant.
     */
    constructor(appId, certificateThumbprint, certificatePrivateKey, tenantId) {
        super();
        (0, assert_1.ok)(appId === null || appId === void 0 ? void 0 : appId.trim(), 'CertificateServiceClientCredentialsFactory.constructor(): missing appId.');
        (0, assert_1.ok)(certificateThumbprint === null || certificateThumbprint === void 0 ? void 0 : certificateThumbprint.trim(), 'CertificateServiceClientCredentialsFactory.constructor(): missing certificateThumbprint.');
        (0, assert_1.ok)(certificatePrivateKey === null || certificatePrivateKey === void 0 ? void 0 : certificatePrivateKey.trim(), 'CertificateServiceClientCredentialsFactory.constructor(): missing certificatePrivateKey.');
        this.appId = appId;
        this.certificateThumbprint = certificateThumbprint;
        this.certificatePrivateKey = certificatePrivateKey;
        this.tenantId = tenantId;
    }
    /**
     * @inheritdoc
     */
    isValidAppId(appId) {
        return __awaiter(this, void 0, void 0, function* () {
            return appId === this.appId;
        });
    }
    /**
     * @inheritdoc
     */
    isAuthenticationDisabled() {
        return __awaiter(this, void 0, void 0, function* () {
            // Auth is always enabled for Certificate.
            return;
        });
    }
    /**
     * @inheritdoc
     */
    createCredentials(appId, audience) {
        return __awaiter(this, void 0, void 0, function* () {
            (0, assert_1.ok)(yield this.isValidAppId(appId), 'CertificateServiceClientCredentialsFactory.createCredentials(): Invalid Managed ID.');
            return new certificateAppCredentials_1.CertificateAppCredentials(this.appId, this.certificateThumbprint, this.certificatePrivateKey, this.tenantId, audience);
        });
    }
}
exports.CertificateServiceClientCredentialsFactory = CertificateServiceClientCredentialsFactory;
